File: java-workshop/part-2-light-mocks/src/main/java/domain/Order.java
package domain;

import java.util.List;
public record Order(
    String id,
    String customerId,
    List<OrderItem> items,
    OrderStatus status,
    double totalPrice,
    String transactionId,
    String trackingNumber,
    DiscountCode discountCode
) {
    public Order {
        items = List.copyOf(items);
    }

    public Order(
        String id,
        String customerId,
        List<OrderItem> items,
        OrderStatus status,
        double totalPrice,
        String transactionId,
        String trackingNumber
    ) {
        this(id, customerId, items, status, totalPrice, transactionId, trackingNumber, null);
    }

    public Order confirm(double totalPrice) {
        return new Order(id, customerId, items, OrderStatus.CONFIRMED, totalPrice, transactionId, trackingNumber, discountCode);
    }

    public Order reject() {
        return new Order(id, customerId, items, OrderStatus.REJECTED, totalPrice, transactionId, trackingNumber, discountCode);
    }
}

File: typescript-workshop/part-2-light-mocks/src/domain/Order.ts
import { OrderItem } from './OrderItem';
import { OrderStatus } from './OrderStatus';

export class Order {
  constructor(
    readonly id: string,
    readonly customerId: string,
    readonly items: ReadonlyArray<OrderItem>,
    readonly status: OrderStatus,
    readonly totalPrice: number,
    readonly transactionId: string | null,
    readonly trackingNumber: string | null,
  ) {}

  confirm(totalPrice: number): Order {
    return new Order(
      this.id,
      this.customerId,
      this.items,
      OrderStatus.Confirmed,
      totalPrice,
      this.transactionId,
      this.trackingNumber,
    );
  }

  reject(): Order {
    return new Order(
      this.id,
      this.customerId,
      this.items,
      OrderStatus.Rejected,
      this.totalPrice,
      this.transactionId,
      this.trackingNumber,
    );
  }
}
